# Notebook Instructions

1. All the <u>code and data files</u> used in this course are available in the downloadable unit of the <u>last section of this course</u>.
2. You can run the notebook document sequentially (one cell at a time) by pressing **shift + enter**. 
3. While a cell is running, a [*] is shown on the left. After the cell is run, the output will appear on the next line.

This course is based on specific versions of Python packages. You can find the details of the packages in <a href='https://quantra.quantinsti.com/quantra-notebook' target="_blank" >this manual</a>.

## DataFrame and Basic Functionality

The underlying idea of a dataframe is based on 'spreadsheets'. In other words, dataframes store data in discrete rows and columns. There are multiple columns in a dataframe. Each column in a dataframe can also have a name. 

Python has a library, `pandas`, which has a very powerful data structure that can help you manage data very effectively. This data structure is called **DataFrame**. 

In this notebook, you will perform the following tasks.

1. [Create a Dataframe](#create-dataframe)
2. [Load the Financial Market Data](#load-data)
3. [Structure of a Dataframe](#structure)
4. [Head and Tail of a Dataframe](#head-and-tail)
5. [Access Elements of a Dataframe](#accessing-elements)
  <br>a. [loc](#loc)
  <br>b. [iloc](#iloc)
  <br>c. [Boolean Indexing](#boolean)
6. [Access a Column in a Dataframe](#access-column)
7. [Remove Rows & Columns from a Dataframe](#remove-rows-columns)
8. [Add Columns to a Dataframe](#add-columns)

<a id='create-dataframe'></a> 
## Create a Dataframe

To create a dataframe, you will first need to import `pandas` library. After importing the library, you can use the `DataFrame` method to create the dataframe. The constructor for a dataframe is as below.

```python
import pandas as pd
pd.DataFrame(data=None, index=None)
```

Let us look at the following example.

In [1]:
# Import pandas as alias pd
import pandas as pd

# A dataframe has a row and a column index
# Data to be stored in the dataframe
my_portfolio = {
    "stock_name": ["Coca Cola", "Facebook", "Apple", "Tesla", "Infosys"],
    "quantity_owned": [1564, 6546, 5464, 6513, 4155],
    "average_buy_price": [950, 160, 120, 270, 15]
}

# Pass 'my_portfolio' as the argument in the dataframe constructor
my_portfolio_frame = pd.DataFrame(my_portfolio)

# Display the dataframe
my_portfolio_frame

stock_name  quantity_owned  average_buy_price
0  Coca Cola            1564                950
1   Facebook            6546                160
2      Apple            5464                120
3      Tesla            6513                270
4    Infosys            4155                 15

<a id='structure'></a> 
## Structure of a Dataframe

Let's look at the output above. `stock_name`, `quantity_owned` and `average_buy_price` are the **columns** of the dataframe. Each column can be accessed by the name of the column. The first column that you see (numbers 0 to 4) is the **index** of the dataframe. Each **row** can be accessed by the index of the row. 

We will come back to this a bit later and see how to access the rows and columns. Before that let's look at the index. Right now, if you want to access the information for any particular stock, you will have to know the row number. Wouldn't it be great, if the stock names are the index instead? You can do this by using the `set_index` method.

Syntax:
```python
DataFrame.set_index(column_name)
```
**column_name:** Name of the column you want to set the index 

In [2]:
# Set the 'stock_name' as the index of the dataframe
my_portfolio_frame = my_portfolio_frame.set_index('stock_name')

# Display the dataframe
my_portfolio_frame.head()

quantity_owned  average_buy_price
stock_name                                   
Coca Cola             1564                950
Facebook              6546                160
Apple                 5464                120
Tesla                 6513                270
Infosys               4155                 15

You can see that now the **stock_name** is the index of the dataframe. Hence, if you want to access the information of any particular asset, you can simply do it by its name.

<a id='load-data'></a>
## Load the Financial Market Data

You can also store the financial market data in a dataframe. These data might be available in form of an excel spreadsheet or a CSV file. You can work with these files by reading them as a dataframe. You can also import data from web sources using different Python packages. This is later covered in detail in the next notebook on importing data. 

For now, the stock price data of Coca Cola from January 2019 to March 2021 is available in a CSV file. You can find this data in the downloadable section at the end of the course.

You can use `pandas.read_csv()` to read the csv file.

Syntax:
```python
pd.read_csv(filename,index_col)
```
1. **filename**: Complete path of the file in the string format
2. **index_col**: The column number to set as index

The data file is stored in a directory `data_modules`. We will store this in a variable path to simplify the code.

In [3]:
# The data is stored in the directory 'data_modules'
path = '../data_modules/'

# Loading data using CSV file
coca_cola = pd.read_csv(path + 'coca_cola_price.csv', index_col=0)

Here, the filename `path + 'coca_cola_price.csv'` is addition of two strings. It becomes `'../data_modules/coca_cola_price.csv'`.

In [4]:
# Display the entire frame
coca_cola

Open       High        Low      Close  Adj Close    Volume
Date                                                                       
2019-01-02  46.939999  47.220001  46.560001  46.930000  43.620419  11603700
2019-01-03  46.820000  47.369999  46.529999  46.639999  43.350868  14714400
2019-01-04  46.750000  47.570000  46.639999  47.570000  44.215286  13013700
2019-01-07  47.570000  47.750000  46.900002  46.950001  43.639008  13135500
2019-01-08  47.250000  47.570000  47.040001  47.480000  44.131630  15420700
...               ...        ...        ...        ...        ...       ...
2021-03-24  51.169998  51.889999  51.150002  51.520000  51.520000  14997400
2021-03-25  51.700001  52.060001  51.169998  52.020000  52.020000  17091900
2021-03-26  52.139999  53.110001  51.759998  53.040001  53.040001  17125100
2021-03-29  52.730000  53.939999  52.709999  53.849998  53.849998  17514100
2021-03-30  53.750000  53.790001  52.930000  53.150002  53.150002  14871800

[565 rows x 6 columns]

From the above output, you can see that:
* The dataframe has 565 rows and 6 columns. 
* The date range for the available data is `2019-01-02` to `2021-03-30`. 
* The data for `Open`, `High`, `Low`, `Close`, `Adj Close`, and `Volume` is available under the respective columns.

<a id='head-and-tail'></a>
## Head and Tail of a Dataframe

You can display the first few rows of the dataframe by using the `head()`. 

Syntax:
```python
DataFrame.head(n)
```
**n**: Number of rows. If not specified, it is 5.

Similarly, you can display the last few rows of the dataframe by using the `tail()`. 

Syntax:
```python
DataFrame.tail(n)
```
**n**: Number of rows. If not specified, it is 5.

Visualising the head or tail of the dataframe will give you an idea of what kind of data you are dealing with.

In [5]:
# Display the head
# You will see the top 5 rows
coca_cola.head()

Open       High        Low      Close  Adj Close    Volume
Date                                                                       
2019-01-02  46.939999  47.220001  46.560001  46.930000  43.620419  11603700
2019-01-03  46.820000  47.369999  46.529999  46.639999  43.350868  14714400
2019-01-04  46.750000  47.570000  46.639999  47.570000  44.215286  13013700
2019-01-07  47.570000  47.750000  46.900002  46.950001  43.639008  13135500
2019-01-08  47.250000  47.570000  47.040001  47.480000  44.131630  15420700

In [6]:
# Display the tail
# You will see the bottom 5 rows
coca_cola.tail()

Open       High        Low      Close  Adj Close    Volume
Date                                                                       
2021-03-24  51.169998  51.889999  51.150002  51.520000  51.520000  14997400
2021-03-25  51.700001  52.060001  51.169998  52.020000  52.020000  17091900
2021-03-26  52.139999  53.110001  51.759998  53.040001  53.040001  17125100
2021-03-29  52.730000  53.939999  52.709999  53.849998  53.849998  17514100
2021-03-30  53.750000  53.790001  52.930000  53.150002  53.150002  14871800

Let's now see how can you access the elements of the dataframe.

<a id='accessing-elements'></a> 
## Access Elements of a Dataframe

<a id='loc'></a> 
### loc

Remember the dataframe `my_portfolio_frame`. 

In [7]:
my_portfolio_frame

quantity_owned  average_buy_price
stock_name                                   
Coca Cola             1564                950
Facebook              6546                160
Apple                 5464                120
Tesla                 6513                270
Infosys               4155                 15

You want only to see the row with the values for Coca Cola. The index of the row is `Coca Cola`. You can access the row of the dataframe using `loc`. 

Syntax:
```python
DataFrame.loc[index]
```

In [8]:
# Display the Coca Cola row
my_portfolio_frame.loc['Coca Cola']

quantity_owned       1564
average_buy_price     950
Name: Coca Cola, dtype: int64

`loc` can also be used to access multiple rows. Let's see the data for all the stocks from Coca Cola to Apple.

Syntax:
```python
DataFrame.loc[start index: end index]
```
1. **start index**: Index of the first row where you want to start.
2. **stop index**: Index of the last row where you want to stop.

Both, start index and stop index are included in the range.

In [9]:
# Display rows from Coca Cola to Apple
my_portfolio_frame.loc['Coca Cola':'Apple']

quantity_owned  average_buy_price
stock_name                                   
Coca Cola             1564                950
Facebook              6546                160
Apple                 5464                120

<a id='iloc'></a> 
### iloc

You can also access the row of a dataframe by its position (row number). This can be done using `iloc`.
Syntax:
```python
DataFrame.iloc[index]
```

Let's continue with the same example. You want to see the details of Coca Cola. It is the first row of the dataframe. Hence the index will be zero. (**Note:** Index always begins with 0.)

In [10]:
# Display first row
my_portfolio_frame.iloc[0]

quantity_owned       1564
average_buy_price     950
Name: Coca Cola, dtype: int64

Like `loc`, `iloc` can also be used to access multiple rows. But there is one difference. The end index is not included in the range.

Syntax:
```python
DataFrame.iloc[start index: end index]
```
1. **start index**: Index of the first row where you want to start. The element at this index is included.
2. **stop index**: Index of the last row where you want to stop. The element at this index is not included.


In [11]:
# Display rows from index 0 to 2
my_portfolio_frame.iloc[0:3]

quantity_owned  average_buy_price
stock_name                                   
Coca Cola             1564                950
Facebook              6546                160
Apple                 5464                120

<a id='boolean'></a> 
### Boolean Indexing

Let's suppose you want to see the stocks for which `quantity_owned` is greater than 5000. How will you do this?

In simple English, you want to filter the dataframe in such a way that `quantity_owned > 5000` is always true. In the previous notebook, you learnt about booleans. Booleans are either `True` or `False`. You can filter your dataframe by indexing it with boolean values. 

Syntax:
```python
DataFrame[boolean filter]
```
**boolean filter:** It is the condition that will always hold true.

In [12]:
# Create a filter such that 'quantity_owned' is greater than 5000
my_portfolio_frame[my_portfolio_frame.quantity_owned > 5000]

quantity_owned  average_buy_price
stock_name                                   
Facebook              6546                160
Apple                 5464                120
Tesla                 6513                270

<a id='access-column'></a> 
## Access a Column in a Dataframe

You can access the column(s) of the dataframe by their name. Look at the example below.

In [13]:
# Column name is passed within quotes within square brackets
my_portfolio_frame[["quantity_owned"]].head()

quantity_owned
stock_name                
Coca Cola             1564
Facebook              6546
Apple                 5464
Tesla                 6513
Infosys               4155

Let us access the quantity owned for Coca Cola.

In [14]:
# Quantity owned for Coca Cola
my_portfolio_frame.loc['Coca Cola', "quantity_owned"]

1564

**Can you guess how will you access the quantity owned for all the assets from Coca Cola to Apple?**

In [15]:
my_portfolio_frame.loc['Coca Cola' : 'Apple', ["quantity_owned"]]

quantity_owned
stock_name                
Coca Cola             1564
Facebook              6546
Apple                 5464

<a id='remove-rows-columns'></a> 
## Remove Rows & Columns from a Dataframe

In the above Coca Cola stock data, it is not necessary that you need all the columns which are present in the .csv file. Hence, to make your dataframe more understandable, you may drop the columns that you do not need using drop function.<br>
<br>
Syntax for dropping rows:<br>
<pre>DataFrame.drop(DataFrame.index[[x,y,z...]])</pre>
where x,y,z are row index values.<br><br>
Syntax for dropping columns:<br>
<pre>DataFrame.drop(['Column_name'])</pre>


In [16]:
# Dropping rows: 2019-01-04, 2019-01-05
# 2019-01-04 is at index 1 and 2019-01-05 is at index 2
coca_cola.drop(coca_cola.index[[1, 2]]).head()

Open       High        Low      Close  Adj Close    Volume
Date                                                                       
2019-01-02  46.939999  47.220001  46.560001  46.930000  43.620419  11603700
2019-01-07  47.570000  47.750000  46.900002  46.950001  43.639008  13135500
2019-01-08  47.250000  47.570000  47.040001  47.480000  44.131630  15420700
2019-01-09  47.080002  47.279999  46.430000  46.570000  43.285809  18032500
2019-01-10  46.660000  47.090000  46.340000  47.070000  43.750546  15976600

In [17]:
# Drop columns 'High' and 'Low'
# axis=1 represents the column axis
coca_cola = coca_cola.drop(['High', 'Low'], axis=1)
coca_cola.head()

Open      Close  Adj Close    Volume
Date                                                 
2019-01-02  46.939999  46.930000  43.620419  11603700
2019-01-03  46.820000  46.639999  43.350868  14714400
2019-01-04  46.750000  47.570000  44.215286  13013700
2019-01-07  47.570000  46.950001  43.639008  13135500
2019-01-08  47.250000  47.480000  44.131630  15420700

<a id='add-columns'></a> 
## Add Columns to a Dataframe

There are multiple ways to add a new column to a dataframe. Let's look at one of the simplest ways. We will calculate the 10-period moving average and store it in a new column, `SMA`.

In [18]:
# Add the 10-period simple moving average of the close price
coca_cola['SMA'] = coca_cola['Close'].rolling(window=10).mean()
coca_cola.tail()

Open      Close  Adj Close    Volume     SMA
Date                                                         
2021-03-24  51.169998  51.520000  51.520000  14997400  51.002
2021-03-25  51.700001  52.020000  52.020000  17091900  51.116
2021-03-26  52.139999  53.040001  53.040001  17125100  51.384
2021-03-29  52.730000  53.849998  53.849998  17514100  51.666
2021-03-30  53.750000  53.150002  53.150002  14871800  51.859

## Conclusion

You have seen some basic functionality of dataframes here which is further used in the course. It has much more to offer. In the next notebook, you will learn to import data from different sources. <br><br>